waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/PanopticOperationHandler.java

```diff
@@ -15,30 +15,82 @@
  */
 package com.hotels.bdp.waggledance.mapping.service;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
 import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
 
 import org.apache.hadoop.hive.metastore.HiveMetaStore.HMSHandler;
 import org.apache.hadoop.hive.metastore.api.TableMeta;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hotels.bdp.waggledance.mapping.model.DatabaseMapping;
+import com.hotels.bdp.waggledance.mapping.service.requests.GetAllDatabasesByPatternRequest;
+import com.hotels.bdp.waggledance.mapping.service.requests.GetAllDatabasesRequest;
+import com.hotels.bdp.waggledance.mapping.service.requests.GetTableMetaRequest;
+import com.hotels.bdp.waggledance.mapping.service.requests.SetUgiRequest;
 
 /**
  * Class responsible for handling the Hive operations that need to combine results from multiple Hive Metastores
  */
-public interface PanopticOperationHandler {
+public abstract class PanopticOperationHandler {
+
+  protected static final long GET_DATABASES_TIMEOUT = TimeUnit.MILLISECONDS.toMillis(8000L);
+  private static final Logger LOG = LoggerFactory.getLogger(PanopticOperationHandler.class);
+  private static final String INTERRUPTED_MESSAGE = "Execution was interrupted: ";
+  private static final String SLOW_METASTORE_MESSAGE = "Metastore {} was slow to respond so results are omitted";
+  private static final long SET_UGI_TIMEOUT = TimeUnit.MILLISECONDS.toMillis(8000L);
+  private static final long GET_TABLE_META_TIMEOUT = TimeUnit.MILLISECONDS.toMillis(400L);
 
   /**
    * Implements {@link HMSHandler#get_all_databases()} over multiple metastores
    *
    * @return list of all databases
    */
-  List<String> getAllDatabases();
+  public abstract List<String> getAllDatabases();
 
   /**
    * Implements {@link HMSHandler#get_database(String)} over multiple metastores
    *
    * @param databasePattern pattern to match
    * @return list of all databases that match the passed pattern
    */
-  List<String> getAllDatabases(String databasePattern);
+  public abstract List<String> getAllDatabases(String databasePattern);
+
+  protected List<String> getAllDatabases(
+      Map<DatabaseMapping, String> databaseMappingsForPattern,
```

Comment:
There's quite a lot of repeated code here for each method, we should try reduce this if possible.

Reply:
I would like to, but from `getAllDatabases` to `setUgi`, each time there is a for loop, a different type of object is created and it takes different parameters.

There is a lot of repetition from `getDatabasesFromFuture` to `getTableMetaFromFuture`, where the only difference is the object that is passed (all the Request classes). But I couldn't find a way to make the function work with generics :/

Reply:
@massdosage: @AnanaMJ and me had a go to tackle some of the duplication. Should be a bit better now.

Reply:
Looks better now, yes!